src/java.base/windows/classes/sun/nio/fs/WindowsPath.java

```diff
@@ -174,38 +174,48 @@ String getPathForPermissionCheck() {
     // use this path for Win32 calls
     // This method will prefix long paths with \\?\ or \\?\UNC as required.
     String getPathForWin32Calls() throws WindowsException {
+        return getPathForWin32Calls(false);
+    }
+
+    String getPathWithPrefixForWin32Calls() throws WindowsException {
+        return getPathForWin32Calls(true);
+    }
+
+    private String getPathForWin32Calls(boolean forceLongPrefix) throws WindowsException {
         // short absolute paths can be used directly
         if (isAbsolute() && path.length() <= MAX_PATH)
-            return path;
-
-        // return cached values if available
-        WeakReference<String> ref = pathForWin32Calls;
-        String resolved = (ref != null) ? ref.get() : null;
-        if (resolved != null) {
-            // Win32 path already available
-            return resolved;
+            return forceLongPrefix ? addPrefix(path) : path;
```

Comment:
Brian and I discussed this issue today and the reason this version didn't work is that it prepends the long path prefix without first using GetFullPathName to resolve the "." and ".." elements in the name. Brian is working on a new version now that will fix this issue and we should end up with a cleaner solution.

Reply:
Committed.